### ISSUE TITLE
Form showing pre-populated data

### ISSUE THREAD

**Issue by ewanduthie** [verbb/formie#1744]
### Question

Not entirely sure how or where to start with diagnosing this. 

A user has reported to us that when they went to submit a multi-step form (for the first time) that the fields were pre-populated with data which was not their own. Curiously, if I try and find a logged submission that contains that data, I can't find one. But the data in the fields is definitely the kind of data we would expect to be submitted through this form. 

Refreshing the page apparently cleared the pre-populated data, but this raises a security/privacy concern.

I'm unable to replicate this, but have been shared a screenshot which verifies the pre-populated data on the form doesn't belong to the person who was attempting to submit for the first time.

The only thing I could think of is someone else was using their machine previously to fill out the form, but that would seem highly unlikely (the pre-populated data belongs to another organisation)

Has anyone else experienced this? What are some steps I can take to try diagnose?

**Comment by engram-design** [comment]
Is there any form of caching - particularly static caching - on the site? 

Do you employ a method of populating the form? It should be impossible to populate another session's submission data into your own session. For a shared computer, this is possible, but the submission data would be removed once logged off, and is only applicable for incomplete submissions. 

Refreshing the page to clear the form seems a bit suspect...

**Comment by scholejo** [comment]
We are experiencing a similar behavior using static caching with Blitz.

We have a multi-step form with ajax enabled on a page that is being cached by Blitz. When my colleague entered his data in that form and went on to the next page of that form, the site was apparently cached including the input values he entered. So afterwards I could see the data he entered when opening the same page on my computer.

After going through the Formie documentation and our configurations I am pretty sure that the reason behind this behavior is that we set the refresh mode of Blitz to "[Clear the cache, regenerate manually or organically](https://putyourlightson.com/plugins/blitz#clear-the-cache-regenerate-manually-or-organically)". This setting causes Blitz to cache a page when it is requested. Formie populates the Form before the page is cached and so the input values get cached alongside the form.

After all, this behavior is pretty obvious and we can fix this by simply excluding pages with forms from caching or by using another refresh mode for the Blitz cache.

Anyway, do you see a possibility to make Forms work together with Blitz' organic refresh mode? Maybe by not populating the form on request but instead along with the CSRF Token? If not, could you please at least add a little hint to your documentation regarding static Caching with Blitz? This could save people some time in the future.

**Comment by engram-design** [comment]
There's only a very specific set of circumstances where a form will be cached with content populated. That's usually when the cache is cleared as the user is carrying out a form submission.

The cache might be empty on page load for a new submission, the page is immediately cached (with no content in the form). That's normal behaviour.

The cache might be empty, and the user has an incomplete submission for their form. They navigate to the page with the form, which will populate that form with their content for their incomplete submission. However, this will add this page to the cache, with their populated content. That's now served to _everyone_.

That's pretty much the only way I see content being cached, particularly for an Ajax form.

I'm not quite sure the best approach to take here. We could simply blow away all incomplete submissions whenever the Blitz cache is cleared, but that seems incredibly heavy-handed. 

Another approach is to populate the form via JavaScript, as opposed to Twig, but that's quite the refactor, and very complex.

I'll have a think about how to address this.

**Comment by scholejo** [comment]
Thanks for the quick reply!

I _think_ I managed to make this work with Blitz' [dynamic content](https://putyourlightson.com/plugins/blitz#dynamic-content) loading.

In my base template I added these inside the `<head>` Tag:

```
<head>
  ...
  {{ craft.formie.renderCss(true) }}
  {{ craft.formie.renderJs(true) }}
</head>
```

Then in my entry template that's extending the base template I added this:

```
      {% set form = entry.form.one() %}
      {{ craft.blitz.includeDynamic(
        '_dynamic/form',
        {
          formId: form.formId,
          formHandle: form.handle,
          prepopulateValues: { someKey: someValue }
        },
        {
          placeholder: '<p>Loading Form...</p>'
        }
      ) }}

      {% js %}
        // Init Forms after Blitz inject
        document.addEventListener('afterBlitzInject', function (event) {
        Formie.initForms();
        });
      {% endjs %}
```

And then in the `_dynamic/form` template I do this:

```
{% set form = craft.formie.forms().handle(formHandle).one() %}
{% if prepopulateValues is defined and prepopulateValues is not empty %}
  {% do craft.formie.populateFormValues(form, prepopulateValues) %}
{% endif %}
{{ craft.formie.renderForm(
  form,
  { themeConfig: { form: { attributes: { id: formId }}}}
) }}
```

I haven't tested this extensively yet, but as the form is rendered on each request, there shouldn't be cached values from other people anymore.

I think this might be the way to go, but could you please let me know if you think there are any possible flaws I haven't seen?

**Comment by engram-design** [comment]
Thanks for sending all that though, I'll dive into this a little more as I've not had experience with the dynamic templates. But that does seem like a good approach.

I suppose this is still going to cause some issues if you have server rewrites enabled, as a request won't even hit Craft. In an ideal world, there would be a mechanism to see if the current request was a "caching" request, and to just cache the empty form, but I don't know how that would be possible - it might!

I'll investigate this further.

**Comment by bencroker** [comment]
> I suppose this is still going to cause some issues if you have server rewrites enabled, as a request won't even hit Craft.

Just an FYI that unlike _cached_ includes, _dynamic_ includes will _always_ hit the server, and _never_ be cached.

Note also that dynamic includes take place over AJAX, and that browsers don’t execute JS returned in XHR responses, in case that plays a part.

**Comment by scholejo** [comment]
> In an ideal world, there would be a mechanism to see if the current request was a "caching" request, and to just cache the empty form, but I don't know how that would be possible - it might!

Maybe Blitz is sending a specific User Agent that you could check against. But that would be a bit ugly I think.
I think this isn't really necessary though. When Blitz is configured to warm the cache actively via a Queue Job, there shouldn't be a session related to any real person when caching. So in that case the problems described earlier should never appear. We decided to not use Blitz in that way though, because we experienced big performance issues with that configuration on less powerful servers. (Blitz pushes A LOT of jobs into the queue after clearing the cache.)

Regarding my code example from above, I had to make some changes to that because the library for the phone field wouldn't be loaded that way. I removed the global JS and CSS from the <head> Tag and now load the form specific JS and CSS directly in the entry template:

```
      {% set form = entry.form.one() %}
      {{ craft.blitz.includeDynamic(
        '_dynamic/form',
        {
          formId: form.formId,
          formHandle: form.handle
          prepopulateValues: { someKey: someValue }
        },
        {
          placeholder: '<p>Loading Form...</p>'
        }
      ) }}

      {{ craft.formie.renderFormCss(form) }}
      {{ craft.formie.renderFormJs(form) }}
      {% js %}
        // Init Forms after Blitz inject
        document.addEventListener('afterBlitzInject', function (event) {
        Formie.initForms();
        });
      {% endjs %}
```

 And I removed the themeConfig part from the dynamic Template:

```
{% set form = craft.formie.forms().handle(formHandle).one() %}
{% if prepopulateValues is defined and prepopulateValues is not empty %}
  {% do craft.formie.populateFormValues(form, prepopulateValues) %}
{% endif %}
{{ craft.formie.renderForm(
  form
) }}
```

This seems to work quite well but I noticed a bit of a strange behavior. With this code, the form never gets prepopulated with previously entered data. Not even my own. I find this acceptable, but when I navigate to a form page with a summary field in it, that field gets populated with the data I have entered in a previous session. This makes me think that even though I'm presented with an empty form, there is an incomplete submission of my own in the background which might get merged with my newly entered data.

Is there maybe an easy way to fetch an incomplete submission of the current user and actively prepopulating the form via populateFormValues()?

**Comment by scholejo** [comment]
Just a quick addition: By adding `{% do craft.formie.setCurrentSubmission(form, null) %}` before the prepopulate part in the dynamic template the form can be rendered without any submission. That way I get a completely fresh form on every request and don't have to worry about mixing my new submission with an old incomplete one.

**Comment by engram-design** [comment]
I was just about to suggest using `{% do craft.formie.setCurrentSubmission(form, null) %}` to override any saved session.

But as I mentioned, going to look at this properly as it's all a little bit _too_ involved to make things work well with Blitz (with any/all configuration) that I'd rather not leave it up to the user like this going forward.

**Comment by scholejo** [comment]
Yeah, I had to put quite the effort into finding this solution. Of course it would be nice to have this work more out-of-the-box or with just little configuration needed. I hope my insights can help here.

**Comment by jornwildenbeest** [comment]
hi @engram-design, 

we are experiencing the same issues as scholejo did.
Using blitz caching and formie with a multipage form, it sometimes happens that a random users fills out the first page of the form, then the page get's cached and everybody see's the cached form with the user data.

Or is sometimes happens that a users is shown page 2 of the form, because that page just happend to be shown when the page was cached by blitz.

Do you have any fixes for this or should we try it with the dynamic content loading?

**Comment by engram-design** [comment]
It's pretty much the nature of static caching as it's not going to hit the server every time to fire the Twig to populate the form. And more importantly, when it does, it's going to then cache that form with those values for _everyone_.

Your options are:
- Don't cache the page with a populated form on it
- Use [dynamic content](https://github.com/verbb/formie/issues/1744#issuecomment-2047387794)

Related to https://github.com/verbb/formie/issues/2001

**Comment by engram-design** [comment]
This sort of handling has been revamped in [4.0.0-beta.1](https://github.com/verbb/formie/releases/tag/4.0.0-beta.1) to better support things. 

But the core issue does still stands - if you statically cache your site, you'll very likely want to disable the new "Restore In-Progress Submissions Automatically" form setting. This essentially disables the automatic "save the users submission on the form" as that's causing the issue in this thread, and is impossible to have work correctly on a static cached site.
